yt/data_objects/data_containers.py

```diff
@@ -2179,3 +2182,93 @@ def _reconstruct_object(*args, **kwargs):
+    the gas fields of the cell at some random location within the cell's volume.
+
+    By default the fields that are created are mass, density, smoothing_length,
+    and the position and velocity fields.
+
+    Parameters
+    ----------
+    ds : Dataset
+        Grid-based dataset used as source
+
+    n_samples : int
+        Number of times to sample the dataset
+
+    fields : list
+        List of additional fields you want to include in returned dataset
+
+    n_neighbors : int
+        Number of neighbors for each node in kdtree to determine
+        smoothing_length
+
+    Returns
+    -------
+    StreamParticlesDataset
+
+    Examples
+    --------
+    >>> ds_grid = yt.load('IsolatedGalaxy/galaxy0030/galaxy0030')
+    >>> ds_sph = monte_carlo_sample(ds_grid)
+    """
+    from yt.frontends.stream.api import \
+        load_particles
+    l_unit = 'code_length'
+    m_unit = 'code_mass'
+    d_unit = 'code_mass/code_length**3'
+    v_unit = 'code_length/code_time'
+    rands = np.random.rand(n_samples)
+    m_tot = ds.r[('gas', 'cell_mass')].sum()
+    bins = np.cumsum(ds.r[('gas', 'cell_mass')] / m_tot)
+    indices = np.digitize(rands, bins)
+    x_off = np.random.uniform(-0.5, 0.5, n_samples)
+    y_off = np.random.uniform(-0.5, 0.5, n_samples)
+    z_off = np.random.uniform(-0.5, 0.5, n_samples)
+    x_pos = ds.r[('gas', 'x')][indices] + x_off*ds.r[('gas', 'dx')][indices]
+    y_pos = ds.r[('gas', 'y')][indices] + y_off*ds.r[('gas', 'dy')][indices]
+    z_pos = ds.r[('gas', 'z')][indices] + z_off*ds.r[('gas', 'dz')][indices]
+    pos = ds.arr([x_pos, y_pos, z_pos]).T.to(l_unit).v
+    masses = (m_tot / n_samples) * np.ones(n_samples)
+
+    # Get smoothing length
+    left_edge = ds.domain_left_edge.to(l_unit).v
+    right_edge = ds.domain_right_edge.to(l_unit).v
+    kdtree = PyKDTree(
+        pos.astype('float64'),
+        left_edge=left_edge,
+        right_edge=right_edge,
+        periodic=ds.periodicity,
+        leafsize=2*int(n_neighbors),
+    )
+    pos_kd = pos[kdtree.idx]
+    hsml = generate_smoothing_length(pos_kd, kdtree, n_neighbors)
+    hsml = hsml[np.argsort(kdtree.idx)]
+
+    data = {'particle_position_x' : (x_pos.to(l_unit), l_unit),
+            'particle_position_y' : (y_pos.to(l_unit), l_unit),
+            'particle_position_z' : (z_pos.to(l_unit), l_unit),
+            'particle_velocity_x' : (ds.r[('gas', 'velocity_x')][indices].to(v_unit), v_unit),
+            'particle_velocity_y' : (ds.r[('gas', 'velocity_y')][indices].to(v_unit), v_unit),
+            'particle_velocity_z' : (ds.r[('gas', 'velocity_z')][indices].to(v_unit), v_unit),
+            'particle_mass' : (masses.to(m_unit), m_unit),
+            'density' : (ds.r[('gas', 'density')][indices].to(d_unit), d_unit),
+            'smoothing_length' : (hsml, l_unit)}
+
+    # Collect dataset code units
+    code_units = {}
+    for key in ['length', 'mass', 'time', 'velocity', 'magnetic']:
```

Comment:
Could this be gotten from the unit registry, instead of via `getattr` and `hasattr`?

Reply:
Yeah, probably.  I stole this from @qobilidop 's PR, since he did such a great job marking this up for his PR #2186 :). :)

Reply:
I wasn't clear where to get the code units of a dataset when I wrote this. I'll look into the unit registry. Thanks for the suggestion!

Reply:
I've switched to use `ds.quan(1, 'code_length').in_cgs()` and the like in my PR. I haven't figured out how to interact with unit registry effectively. Is this usage okay?

Reply:
I'm not an expert with the unit_registry, so @ngoldbaum would be the person to answer that question.